File: frontend/lib/types/thread.ts
// #766 (ADR 0027): the runner's coach threads — the switcher list shape, the
// single-thread read, and the turn-send payload. Mirrors backend
// app/schemas/thread.py.

import { ChatMessage } from "./chat";

export interface ThreadAnchor {
  activity_id: string;
  name?: string | null;
  start_date?: string | null;
  distance_m?: number | null;
  type?: string | null;
}

export interface ThreadListItem {
  id: string;
  // Server-resolved display title: the written/runner title when set, else the
  // first user message trimmed. Never empty for a thread with a turn in it.
  title: string;
  snippet?: string | null;
  last_message_at?: string | null;
  anchor?: ThreadAnchor | null;
}

export interface ThreadDetail {
  id: string;
  title: string;
  created_at: string;
  last_message_at?: string | null;
  anchor?: ThreadAnchor | null;
  messages: ChatMessage[];
}

/** One session on an amendment's card, as the runner would have it (#987).
 *  Display-only and idless: nothing here is a row yet. */
export interface ProposedSessionRow {
  date: string;
  title: string;
  intent: string;
  discipline: string;
  distance_m?: number | null;
  duration_s?: number | null;
  /** New to this day, so the sheet marks it. */
  changed: boolean;
}

export interface ProposedActionFrame {
  action_type:
    | "check_in"
    | "intent"
    | "split_block"
    | "merge_blocks"
    | "complete_session"
    | "adjust_session"
    | "draft_plan"
    | "amend_plan";
  token: string;
  description: string;
  confirm_label: string;
  dismiss_label: string;
  /** What would actually change, in the words the record uses afterwards.
   *  Server-derived from the settled amendment rather than from what was asked
   *  for, which is the point: a rewrite that drops a session the runner meant to
   *  keep says so here, before they agree to it. */
  changes?: string[];
  /** The whole window as it would stand. Only `amend_plan` fills it. */
  week?: ProposedSessionRow[];
}

/** What the server did, echoed back on confirm. `message` is present only when
 *  the write lands somewhere the runner is not looking — a plan is drafted on
 *  the worker, so the card vanishing is otherwise the only feedback there is. */
export interface ProposedActionResult {
  action_type: string;
  message?: string | null;
  /** What an amendment actually did, computed from the rows it wrote. Reported
   *  because it is the one thing that must be able to disagree with the card. */
  changes?: string[];
}

export interface ThreadMessageSend {
  message: string;
  thread_id?: string;
  anchor_activity_id?: string;
  asked_from?: string;
}

File: backend/app/schemas/thread.py
"""Thread read/write schemas (#766, ADR 0027).

The switcher list, the single-thread read, rename, and the thread-turn send.
A thread's display title resolves server-side: the written/runner title when
set, else the first user message trimmed — an untitled thread needs a name the
moment it has one turn in it.
"""

from datetime import datetime
from typing import List, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.chat import ChatMessageRead

# The switcher folds threads idle beyond this many days under "Earlier"
# client-side; the server just orders by recency.
TITLE_MAX_LENGTH = 200
ASKED_FROM_MAX_LENGTH = 64


class ThreadAnchor(BaseModel):
    """The anchored activity, for the switcher chip and the report-at-head
    projection (the report itself is fetched through the existing coach-report
    endpoint — a read-time projection, never a copy)."""

    activity_id: UUID
    name: Optional[str] = None
    start_date: Optional[datetime] = None
    distance_m: Optional[float] = None
    type: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


class ThreadListItem(BaseModel):
    id: UUID
    title: str
    snippet: Optional[str] = None
    last_message_at: Optional[datetime] = None
    anchor: Optional[ThreadAnchor] = None


class ThreadListResponse(BaseModel):
    threads: List[ThreadListItem]


class ThreadDetail(BaseModel):
    id: UUID
    title: str
    created_at: datetime
    last_message_at: Optional[datetime] = None
    anchor: Optional[ThreadAnchor] = None
    messages: List[ChatMessageRead]


class ThreadRename(BaseModel):
    title: str = Field(min_length=1, max_length=TITLE_MAX_LENGTH)


class ThreadMessageSend(BaseModel):
    message: str = Field(min_length=1)
    # Continue this thread; absent = start a new one with this message.
    thread_id: Optional[UUID] = None
    # Anchor for a NEW thread only (ignored when thread_id is given): the
    # activity whose page the thread was born on. A framing hint, never a data
    # boundary (ADR 0027).
    anchor_activity_id: Optional[UUID] = None
    # The label of the screen the turn was asked from (provenance only; the
    # resolved screen view is slice 3, ADR 0028).
    asked_from: Optional[str] = Field(default=None, max_length=ASKED_FROM_MAX_LENGTH)
